### ISSUE TITLE
Extensions get used even if document doesn't match

### ISSUE THREAD

**Issue by AObuchow** [eclipse/lemminx#653]
As mentioned [here](https://github.com/eclipse/lemminx-maven/issues/27), Lemminx-Maven's "participants" (eg. DiagnosticParticipant) seems to be still used even if the document doesn't match. 

Our implementation of `IXMLExtension.match()` looks like:
```Java
	public static boolean match(DOMDocument document) {
		return document.getDocumentURI().endsWith(pom.xml);
	}
```
As a workaround, I'm going to try adding the following to each "participant" to prevent processing a non-pom.xml file:
```Java
if (!MavenPlugin.match(...)) {
  return;
}
```

I haven't looked into how `IXMLExtension.match()` is being used, but I imagine it's used to determine whether or not an extension should be started up or not. Perhaps it should also be used when going through the extension registry to determine if an extension's "participant" function should be called or not.

**Comment by angelozerr** [comment]
I totally agree that a kind of matcher must be provided because we do that for instance to avoid validating a XML file with the XSD validator which validate syntax of XML Schema (xs:element, etc) and should not be applied for a simple XML file. See https://github.com/eclipse/lemminx/blob/d0a94fd0a9a588450e63f9c2b622c90b0b09b743/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/participants/diagnostics/XSDDiagnosticsParticipant.java#L32

I already thought about and my idea was:

 * provide a new interface IDocumentSelector:

```java
package org.eclipse.lemminx.services.extensions;

import org.eclipse.lemminx.dom.DOMDocument;

public interface IDocumentSelector {

	boolean match(DOMDocument document);
}
```

 * implement this interface 
```java
public class MavenDocumentSelector implements IDocumentSelector  {

	public boolean match(DOMDocument document) {
             return "pom.xml".equals(document.getDocumentURI());
        }
}

```
 * use it when you register participant on the `IXMlExtension#start`

```java
registry.registerDiagnosticsParticipant(mavenDocumentSelector, diagnosticsParticipant);
```

**Comment by AObuchow** [comment]
The `IDocumentSelector` seems like a good idea to me. @mickaelistria WDYT?

**Comment by mickaelistria** [comment]
Whatever works ;)
However, I would asusme this DocumentSelector "match" method should actually be part of the IXMLExtension interface, and the participants would register passing the IXMLExtension instance that registers them. With this, we know which extensions the participants come from, so when stopping the extension, Lemminx could detect the participant to detach as well.

**Comment by AObuchow** [comment]
> However, I would asusme this DocumentSelector "match" method should actually be part of the IXMLExtension interface, and the participants would register passing the IXMLExtension instance that registers them. With this, we know which extensions the participants come from, so when stopping the extension, Lemminx could detect the participant to detach as well.

+1 since Match is already defined in the IXMLExtension interface, and this would remove the need for an extension to [unregister participants themselves](https://github.com/eclipse/lemminx-maven/blob/master/lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenPlugin.java#L128)

**Comment by angelozerr** [comment]
> since Match is already defined in the IXMLExtension interface,

No it doesn't define this match method

**Comment by AObuchow** [comment]
> > since Match is already defined in the IXMLExtension interface,
> 
> No it doesn't define this match method

You're right, my mistake. We just happen to have it defined in our class which implements the IXMLExtension interface.
